src/sqlfluff/core/config.py

```diff
@@ -429,15 +438,32 @@ def __init__(
         # Dialect and Template selection.
         # NB: We import here to avoid a circular references.
         from sqlfluff.core.dialects import dialect_selector
-        from sqlfluff.core.templaters import templater_selector
 
         self._configs["core"]["dialect_obj"] = dialect_selector(
             self._configs["core"]["dialect"]
         )
-        self._configs["core"]["templater_obj"] = templater_selector(
+        self._configs["core"]["templater_obj"] = self.get_templater(
             self._configs["core"]["templater"]
         )
 
+    def __getstate__(self):
+        # Copy the object's state from self.__dict__ which contains
+        # all our instance attributes. Always use the dict.copy()
+        # method to avoid modifying the original state.
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state["_plugin_manager"]
+        return state
+
+    def __setstate__(self, state):
+        # Restore instance attributes
+        self.__dict__.update(state)
+        # NB: We don't reinstate the plugin manager, but this should only
+        # be happening between threads where the plugin manager should
+        # probably be fresh in any case.
+        # NOTE: This means that registering user plugins directly is not
+        # thread safe.
```

Comment:
To me, "not thread safe" means it's subject to race conditions. In this case, it's more that they'll be lost in other processes or threads. Maybe clarify this?